### PR TITLE
[common/cache] add a basic lru cache

### LINKED ISSUE #1429: [common] common/cache mod
**Summary**

For LRU caching, memory and disk.

### PR DESCRIPTION
**Pull request by PsiACE**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

- a basic lru cache for memory/disk, derived from sccache/lru_disk_cache

**TODO**

- [x] Reconstructing the interface for memory cache, provide a common abstract interface
- [x] Reconstructing disk cache
- [x] Integrating with datafuse
- [x] Polishing, docs and more

## Changelog

- New Feature
- Improvement

## Related Issues

Fixes #1429 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by ZhiHanZ** [comment]
folly has some interesting LRU cache implementations:
https://github.com/facebook/folly/blob/master/folly/container/EvictingCacheMap.h

**Comment by ZhiHanZ** [comment]
one question about disk based LRU cache, could we assume the size for persistence storage, for example, in sccache, we could set the cache size to 10Gi or more

**Comment by PsiACE** [comment]
> one question about disk based LRU cache, could we assume the size for persistence storage, for example, in sccache, we could set the cache size to 10Gi or more

At present there are no differences between it and the cache in sccache, except for some differences in dependencies and features.

**Comment by BohuTANG** [comment]
Tips:
It would better define the api(the api keeping TODO is ok) first, then let datafuse integrate it, and finally make it workable

File: common/cache/src/disk_cache.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+            })
+        })
+        .collect();
+    // Sort by last-modified-time, so oldest file first.
+    files.sort_by_key(|k| k.0);
+    Box::new(files.into_iter().map(|(_mtime, path, size)| (path, size)))
+}
+
+/// An LRU cache of files on disk.
+pub type LruDiskCache = DiskCache<LruCache<OsString, u64, DefaultHashBuilder, FileSize>>;
```
Comment (PsiACE) [inline common/cache/src/disk_cache.rs:62]: We can easily define another disk cache based on a different policy by simply implementing the `common_cache::cache::Cache` trait.

File: common/cache/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+mod lru_test;
+
+use std::borrow::Borrow;
+use std::hash::BuildHasher;
+use std::hash::Hash;
+
+use crate::Meter;
+
+/// A trait for a cache.
+pub trait Cache<K, V, S, M>
```
Comment (PsiACE) [inline common/cache/src/cache.rs:26]: An almost universal Cache trait

**Review by dantengsky — APPROVED** [review]
Looks great to me, :+1:

**Pull request merged**